packages/runtime/container-runtime/src/opDecompressor.ts

```diff
@@ -0,0 +1,72 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { decompress } from "lz4js";
+import { ISequencedDocumentMessage } from "@fluidframework/protocol-definitions";
+import { assert, IsoBuffer } from "@fluidframework/common-utils";
+import { CompressionAlgorithms } from ".";
+
+/**
+ * State machine that "unrolls" contents of compressed batches of ops after decompressing them.
+ * This class relies on some implicit contracts defined below:
+ * 1. A compressed batch will have its first message with batch metadata set to true and compressed set to true
+ * 2. Messages in the middle of a compressed batch will have neither batch metadata nor the compression property set
+ * 3. The final message of a batch will have batch metadata set to false
+ * 4. An individually compressed op will have undefined batch metadata and compression set to true
+ */
+export class OpDecompressor {
+    private activeBatch = false;
+    private rootMessage: ISequencedDocumentMessage | undefined;
+    private processedCount = 0;
+
+    constructor() {
+
+    }
+
+    public processMessage(message: ISequencedDocumentMessage): ISequencedDocumentMessage {
+        // Beginning of a compressed batch
+        if (message.metadata?.batch === true
+            && (message.metadata?.compressed || (message as any).compression === CompressionAlgorithms.lz4)) {
+            assert(this.activeBatch === false, "shouldn't have multiple active batches");
+
+            this.activeBatch = true;
+            this.rootMessage = message;
```

Comment:
I do not think you need to remember whole message - only content should be sufficient.
In general, we should be careful around holding object that are mutable - further layers in ContainerRuntime can mutate it, so holding a reference to it here is very dangerous.

Reply:
Good call, I've changed it to only hang onto the contents rather than the whole message.